### ISSUE TITLE
Fix me if i wrong

### ISSUE THREAD

**Issue by hackitdown** [NourozR/Stock-Price-Prediction-of-Apple-Inc.-Using-Recurrent-Neural-Network-#2]
1. I think your predictions are delayed for one step and if you zoom in your graphic you'll see this
2. Predictions itself are just a data of the previous step with some lag

i don't think it's correct model
or maybe it's just a oddities of drawing....

![figure_1](https://user-images.githubusercontent.com/28023390/36971084-aeb90490-207b-11e8-9468-0dac107dec96.png)
please fix me if i wrong

**Comment by NourozR** [comment]
If you don't think it's the correct model, then please indicate the line of code where you think there is a problem. Looking at the plot I don't think anyone can ascertain about wrong model - besides, my model has some test errors around 1-1.5 - so some level of mismatch with trained data and test data are highly plausible, to be frank.

**Comment by MiVaVo** [comment]
Hi
I think the problem is not in code it is more fundamental. Your LSTM were not able to understand, what value to predict. It does not know what value will be in the future, so the best gues is prevous value.  It is kind of overfitting in time series analysis. Such prediction is useless ,because your prediction does not provide any insight on what will be, but only predicts what happened, though you showed good visualization on your page and have a low loss. 
How to fix it?
There was not relevant study regarding using LSTM for stock market predictions, though there are lots of articles on web that have the same fundamental problem using LSTM for time series predictions: for example https://machinelearningmastery.com/time-series-prediction-lstm-recurrent-neural-networks-python-keras/

**Comment by NourozR** [comment]
@MiVaVo there was a test set and there is a test accuracy. Did you go through the whole thing? 

If this is useless, please help me to understand the "real" scenario. I will be very grateful. Thanks!

**Comment by MiVaVo** [comment]
Well, rmse is not informative in this case. You rmse can be low and you will still predict future price as previous. I have not seen so far the measure of accuracy that will estimate how good time series is predicted. But instead you can try to compute accuracy of your forecast on test set in terms of binary output. So if your forecast goes up(down) and you predicted up(down) for the next period, it will be correct prediction. In other cases the prediction will be wrong. And finally compute (number of correct predictions)/ (number of correct predictions + number of wrong predictions). 
I do not have time to dig deep into your whole code, though I examined the code the autor I referenced and there were exactly the the same 'symptoms' of the problem, so I suppose the problem is the same.

**Comment by NourozR** [comment]
I'm not convinced, to be frank.

**Comment by janeshin059** [comment]
> Well, rmse is not informative in this case. You rmse can be low and you will still predict future price as previous. I have not seen so far the measure of accuracy that will estimate how good time series is predicted. But instead you can try to compute accuracy of your forecast on test set in terms of binary output. So if your forecast goes up(down) and you predicted up(down) for the next period, it will be correct prediction. In other cases the prediction will be wrong. And finally compute (number of correct predictions)/ (number of correct predictions + number of wrong predictions).
> I do not have time to dig deep into your whole code, though I examined the code the autor I referenced and there were exactly the the same 'symptoms' of the problem, so I suppose the problem is the same.

Hi @MiVaVo, I was searching for the lag problem in LSTM in stock market prediction and I came across your comment. Even though it is really important, the problem that the issuer and you pointed out is not that much highlighted in many articles and studies. 
Your comment was inspiring and I'll try it out, thanks! And for the owner of this repo, thanks for implementing the code and visualizing.

**Comment by Apostatee** [comment]
> Well, rmse is not informative in this case. You rmse can be low and you will still predict future price as previous. I have not seen so far the measure of accuracy that will estimate how good time series is predicted. But instead you can try to compute accuracy of your forecast on test set in terms of binary output. So if your forecast goes up(down) and you predicted up(down) for the next period, it will be correct prediction. In other cases the prediction will be wrong. And finally compute (number of correct predictions)/ (number of correct predictions + number of wrong predictions).
> I do not have time to dig deep into your whole code, though I examined the code the autor I referenced and there were exactly the the same 'symptoms' of the problem, so I suppose the problem is the same.

totally agree,  the author did not fully realize his errors. This repo could mislead others easily.
